source/agora/common/Serializer.d

```diff
@@ -588,27 +595,38 @@ public T deserializeFull (T) (scope const(ubyte)[] data) @safe
 public T deserializeFull (T) (scope DeserializeDg dg,
     const ref DeserializerOptions opts = DeserializerOptions.Default) @safe
 {
-    import geod24.bitblob;
-
     // Custom deserialization trumps everything
     static if (hasFromBinaryFunction!T)
         return T.fromBinary!T(dg, opts);
 
-    // BitBlob are fixed size and thus, value types
-    // If we use an `ubyte[]` overload, the deserializer looks for the length
-    else static if (is(T : BitBlob!N, size_t N))
-        return T(dg(T.Width));
-
     // Static array needs to be handled before arrays
     // Note: This might be optimizable for small types (char, ubyte)
     else static if (is(T : E[N], E, size_t N))
     {
-        E deserializeEntry () ()
+        // Small type optimization
+        static if (!hasFromBinaryFunction!E && isSomeChar!E)
         {
-            return deserializeFull!E(dg, opts);
+            E[] process () @trusted
+            {
+                import std.utf;
+                auto record = cast(E[]) (dg(E.sizeof * length));
+                record.validate();
+                return record;
+            }
+            return process()[0 .. N];
         }
+        // `bytes` and `bool` and qualified
+        else static if (!hasFromBinaryFunction!E && E.sizeof == 1)
+            return (() @trusted { return (cast(E[]) dg(N)); })()[0 .. N];
         // Note: This does not allocate because `staticMap` yields a tuple
-        return [ Repeat!(N, deserializeEntry) ];
+        else
+        {
+            E deserializeEntry () ()
```

Comment:
Doesn't this allocate a closure? Or is the compiler smart enough to optimize it?

Reply:
Compiler smart as far as I recall. See the comment. I would add a `@nogc` test if I could.